platform.local.txt

```diff
@@ -6,12 +6,16 @@ tools.ulptool.path={runtime.tools.ulptool.path}/
 
 ## tool name
 tools.ulptool.cmd=esp32ulp_build_recipe.py
+tools.ulptool.c.cmd=recipe_c_combine_pattern.py
 
 ## ulp build tool
 compiler.s.cmd=python "{tools.ulptool.path}{tools.ulptool.cmd}"
 
-## ulp_main.ld file address
+## ulp_main.ld file address																																																																																												
 compiler.c.elf.extra_flags="-L{build.path}/sketch/" -T ulp_main.ld "{build.path}/sketch/ulp_main.bin.bin.o"
 
 ## Compile s (ulp) files
 recipe.hooks.core.postbuild.01.pattern={compiler.s.cmd} {compiler.cpreprocessor.flags} -b {build.path} -p {runtime.platform.path} -u {compiler.ulp.path} -x {compiler.path} -t {tools.ulptool.path} --DF_CPU={build.f_cpu} --DARDUINO={runtime.ide.version} --DARDUINO_={build.board} --DARDUINO_ARCH_={build.arch} --DARDUINO_BOARD="{build.board}" --DARDUINO_VARIANT="{build.variant}"
+
```

Comment:
I don't think I want to override the esp combine recipe because if they make any changes I have to change the ulptool recipe too. Is there a way for python to edit the **platform.local.txt** **compiler.c.elf.extra_flags** at compile time?

Reply:
I will try.

Reply:
I think to make 2 files and copy the good one to the hardware directory at run time.

Reply:
That might work and make it so the user doesn't have to manually copy **platform.local.txt** over, just do it in the ulp build script?

Reply:
It works but... when you compile the (non) upl for the first time it fail (after it works like a charm). So I think that the 'platform.local.txt' is already in memory thus it does not affect the compilation routine. Second consideration to take in account is the fact that we don't know if any other 'platform.local.txt' file is already used.

For my test I've in modified this par in the recipe script:

```suggestion
    if not ulp_files:
        sys.stdout.write('No ULP Assembly File(s) Detected...\r')
        try:
            with open('ulp_main.ld',"w") as fld: pass
            shutil.copyfile(os.path.join(tpath , 'platform.local_non_ulp.txt'), os.path.join(ppath, 'platform.local.txt'))
        except Exception as error:
            sys.stdout.write(error)
    else:
        shutil.copyfile(os.path.join(tpath , 'platform.local_ulp.txt'), os.path.join(ppath, 'platform.local.txt'))
        build_ulp(bpath, ppath, xpath, upath, tpath, ulp_files, board_options, True)
    sys.exit(0)
```

And the 2 other files are located in the 'ulptool/src/' and are named  'platform.local_ulp.txt' and 'platform.local_non_ulp.txt' with the 'compiler.c.elf.extra_flags' commented.

Reply:
OK, I'm on a 'better' fix. Again we need to compile two time but, here, we just toggle the extra flag expression.

```
def update_platform_local(ppath, enable_extra_flag = True):
    temp = []
    ## Read and evalute the next action (overwrite or return)
    with open(os.path.join(ppath, 'platform.local.txt'), "r") as pltf:
        for line in pltf:
            temp.append(line)

    temp_write = temp

    if 'compiler.c.elf.extra_flags' in ' '.join(temp):
        for index_line in range(len(temp)):      
            if 'compiler.c.elf.extra_flags' in temp[index_line]:
                if (enable_extra_flag and not ('#' in temp[index_line])) or (not enable_extra_flag and ('#' in temp[index_line])):
                    print('Quit')
                    return
                
                elif enable_extra_flag and ('#' in temp[index_line]):       # Update to active ulp compilation
                    print('Active ULP compilation')
                    extra_flag = temp[index_line][temp[index_line].find('compiler.c.elf.extra_flags'):]
                    temp_write.remove(temp[index_line])
                    temp_write.insert(index_line, extra_flag)
                    
                elif not enable_extra_flag and not ('#' in temp[index_line]):   # Update to desactivate the ulp compilation
                    print('Desactive ULP compilatiob')
                    extra_flag = '## ' + temp[index_line][temp[index_line].find('compiler.c.elf.extra_flags'):]
                    temp_write.remove(temp[index_line])
                    temp_write.insert(index_line, extra_flag)
                
    else:
        return "Unable to uptdate 'platform.local.txt'"

    # Overwrite the file
    with open(os.path.join(ppath, 'platform.local.txt'), "w") as pltf:
        for line in temp_write:
            pltf.write(line)
    return
```

And:

```
    if not ulp_files:
        sys.stdout.write('No ULP Assembly File(s) Detected...\r')
        try:
            with open('ulp_main.ld',"w") as fld: pass
            # add # if not present
            update_platform_local(ppath, enable_extra_flag = False)
        except Exception as error:
            sys.stdout.write(error)
    else:
        # remove # if present
        update_platform_local(ppath, enable_extra_flag = True)
        build_ulp(bpath, ppath, xpath, upath, tpath, ulp_files, board_options, True)
    sys.exit(0)
```

I will commit it.